File: solutions/170-two-sum-iii-data-structure-design/two-sum-iii-data-structure-design.ts
class TwoSum {
    num_counts: { [key: number]: number };

    constructor() {
        this.num_counts = {};
    }

    add(number: number): void {
        if (number in this.num_counts) this.num_counts[number]++;
        else this.num_counts[number] = 1;
    }

    find(value: number): boolean {
        for (let num in this.num_counts) {
            const complement = value - Number(num);
            if (complement != Number(num)) {
                if (complement in this.num_counts) return true;
            } else {
                if (this.num_counts[Number(num)] > 1) return true;
            }
        }
        return false;
    }
}

File: solutions/170-two-sum-iii-data-structure-design/two-sum-iii-data-structure-design.go
type TwoSum struct {
    num_counts map[int]int
}

/** Initialize your data structure here. */
func Constructor() TwoSum {
    return TwoSum{num_counts: make(map[int]int)}
}

/** Add the number to an internal data structure..
**/
func (this *TwoSum) Add(number int) {
    if _, ok := this.num_counts[number]; ok {
        this.num_counts[number]++
    } else {
        this.num_counts[number] = 1
    }
}

/** Find if there exists any pair of numbers which sum is equal to the value. */
func (this *TwoSum) Find(value int) bool {
    for num := range this.num_counts {
        complement := value - num
        if complement != num {
            if _, ok := this.num_counts[complement]; ok {
                return true
            }
        } else {
            if this.num_counts[num] > 1 {
                return true
            }
        }
    }
    return false
}